util/tracecpu/collector.go

```diff
@@ -0,0 +1,333 @@
+		ts.mu.Unlock()
+	}
+}
+
+// RegisterNormalizedPlan is like RegisterNormalizedSQL, but for normalized plan strings.
+// TODO: benchmark test concurrent performance
+func (ts *TopSQLCollector) RegisterNormalizedPlan(planDigest string, normalizedPlan string) {
+	ts.planRegisterChan <- &planRegisterJob{
+		planDigest:     planDigest,
+		normalizedPlan: normalizedPlan,
+	}
+}
+
+// this should be the only place where the normalizedPlanMap is set
+func (ts *TopSQLCollector) registerNormalizedPlanWorker(planDecoder planBinaryDecodeFunc) {
+	// NOTE: if use multiple worker goroutine, we should make sure that access to the digest map is thread-safe
+	for {
+		job := <-ts.planRegisterChan
+		ts.mu.RLock()
+		_, exist := ts.normalizedPlanMap[job.planDigest]
+		ts.mu.RUnlock()
+		if exist {
+			continue
+		}
+		planDecoded, err := planDecoder(job.normalizedPlan)
+		if err != nil {
+			fmt.Printf("decode plan failed: %v\n", err)
+			continue
+		}
+		ts.normalizedPlanMap[job.planDigest] = planDecoded
+	}
+}
+
+// snapshot will collect the current snapshot of data for transmission
+// This could run in parallel with `Collect()`, so we should guard it by a mutex.
+func (ts *TopSQLCollector) snapshot() []*tipb.CPUTimeRequestTiDB {
+	ts.mu.RLock()
+	defer ts.mu.RUnlock()
+	total := len(ts.topSQLMap)
+	batch := make([]*tipb.CPUTimeRequestTiDB, total)
+	i := 0
+	for key, value := range ts.topSQLMap {
+		sqlDigest, planDigest := decodeCacheKey(key)
+		normalizedSQL := ts.normalizedSQLMap[sqlDigest]
+		normalizedPlan := ts.normalizedPlanMap[planDigest]
+		batch[i] = &tipb.CPUTimeRequestTiDB{
+			TimestampList:  value.TimestampList,
+			CpuTimeMsList:  value.CPUTimeMsList,
+			SqlDigest:      sqlDigest,
+			NormalizedSql:  normalizedSQL,
+			PlanDigest:     planDigest,
+			NormalizedPlan: normalizedPlan,
+		}
+		i++
+	}
+	return batch
+}
+
+func (ts *TopSQLCollector) sendBatch(stream tipb.TopSQLAgent_CollectCPUTimeClient, batch []*tipb.CPUTimeRequestTiDB) error {
+	for _, req := range batch {
+		if err := stream.Send(req); err != nil {
+			logutil.BgLogger().Error("TopSQL: send stream request failed, %v", zap.Error(err))
+			return err
+		}
+	}
+	// response is Empty, drop it for now
+	_, err := stream.CloseAndRecv()
+	if err != nil {
+		logutil.BgLogger().Error("TopSQL: receive stream response failed, %v", zap.Error(err))
+		return err
+	}
+	return nil
+}
+
+// sendToAgentWorker will send a snapshot to the gRPC endpoint every interval
+func (ts *TopSQLCollector) sendToAgentWorker(interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	for {
+		select {
+		case <-ticker.C:
+			batch := ts.snapshot()
+			sendingTimeout := interval - 10*time.Second
+			if sendingTimeout < 0 {
+				sendingTimeout = interval / 2
+			}
+			conn, stream, _, err := newAgentClient(ts.agentGRPCAddress, sendingTimeout)
+			if err != nil {
+				log.Printf("ERROR: failed to create agent client, %v\n", err)
+				continue
+			}
+			if err := ts.sendBatch(stream, batch); err != nil {
```

Comment:
Maybe worth to try this config in order to reduce streaming cost, which *seems* to batch multiple requests together: https://pkg.go.dev/google.golang.org/grpc#WithWriteBufferSize

I recommend to write a benchmark to see how it performs when sending large amount (e.g. 5000) of stream elements.

Reply:
Seems like it's taking effect here, as a flush threshold: https://github.com/grpc/grpc-go/blob/master/internal/transport/http_util.go#L369-L371

The default write buffer size is `32KB`, I'll leave a note here for benchmark test and parameter tuning.